docs/proposals/201909_thanos_functional_sharding.md

```diff
@@ -0,0 +1,153 @@
+
+Example Sources are: Sidecar, Rule, Thanos Receive.
+
+## Goals
+
+Our goals for this design it to find and implement solution for:
+
+* Sharding browsing metrics from the object storage:
+  * e.g Selecting what blocks Store Gateway should expose.
+* Pre-filtering which shards Querier should touch during query. 
+  * This might mean advertise labels manipulation.
+* Sharding compaction/downsampling of metrics in the object storage. 
+  * NOTE: We need to be really careful to not have 2 compactors working on the same Source. This means careful upgrades/configuration changes or coordination. 
+    
+## No Goals
+
+* Time partitioning
+* "Merging" sources together virtually for downsampling/compactions across single Source that just changes external labels.
+* Bloom filters for app metrics within blocks.
+* Allow all permutations of object storage setups:
+    * User uses multiple object storages for all sources?
+    * User uses single object storage for all sources?
+    * User uses any mix of object storages for sources. They put all in different subdirs/subpaths.
+
+This design is orthogonal for multi-buckets/multi-prefix support. 
+
+## Proposal
+
+On each component that works on the object storage (e.g Store GW and Compactor), add `--selector.relabel-config` (and corresponding `--selector.relabel-config-file`) that will
+be used to filter out what blocks should be selected for operations. Examples:
+
+* We want to run Compactor only for blocks with `external_labels` being `cluster=A`. We will run second Compactor for blocks with `cluster=B` external labels.
+
+Similar to [promtail](https://github.com/grafana/loki/blob/master/docs/promtail.md#scrape-configs) this config will follow native
+[Prometheus relabel-config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config) syntax.
+
+The relabel config will define filtering process done on **every** synchronization with object storage.
+
+We will allow manipulating with several of inputs:
+
+* `__block_id`
+* External labels:
+  * `<name>` 
+* `__block_objstore_bucket_endpoint` 
+* `__block_objstore_bucket_name`
+* `__block_objstore_bucket_path`
+
+Output:
+
+* External labels to be advertised for this block.
+* If output is empty, drop block. 
+
+By default, on empty relabel-config, all external labels are assumed.
+Intuitively blocks without any external labels will be ignored.
+
+Example usages would be:
+
+* Drop blocks which contains external labels cluster=A
+```yaml
+- action: drop
+  regex: "A"
+  source_labels:
+  - cluster
+```
+* Keep only blocks which contains external labels cluster=A
+```yaml
+- action: keep
+  regex: "A"
+  source_labels:
+  - cluster
+```
+* Drop cluster label from external labels for each blocks (if present).
+```yaml
+- action: labeldrop
+  source_labels:
+  - cluster
+```
+* Add `datacenter=ABC` external label to the result.
+```yaml
+- target_label: datacenter
```

Comment:
if we're adding this, this seems redundant with the `--label` flag that most components have, should we still keep those then?

Reply:
This is not going to work easily. This is because we might want to advertise things like 

```
cluster = A | B 
```

And that is definitely not supported by current relabelling.

Reply:
And I think to have proper `advertise` logic we might need to extend relabelling for that